### ISSUE TITLE
Proc=None causes profiling to go into infinite loop

### ISSUE THREAD

**Issue by robertsawko** [erdc-cm/proteus#214]
I am not sure at what point that happened but suddenly my proteus run that use `import proteus.iproteus *` freeze. It turns out that they go into an infinite loop whilst handling `preInitBuffer`. After closer inspection I realized that this is the problem is in `Profiling.py`:

```python
def openLog(filename,level,logLocation=None):
    ....
    for string,level,data in preInitBuffer:
        logEvent(string,level,data)

...
def logEvent(stringIn, level=1, data=None):
...
    elif procID==None:
        preInitBuffer.append((stringIn,level,data))
```
So assuming `preInitBuffer` was not empty it gets bigger and bigger. In my case it starts with `NO PROTEUS_PRELOAD_LIBS LOADED; CHECK config.py`. Also before my `procID` was equal to 0 and now it's 'None' which causes the other if to activate. I was looking around for a while, but couldn't understand the difference so I temporarily commented out the append line. I am afraid of side effects though. Could you please look at it, too?

**Comment by ahmadia** [comment]
@robertsawko - Thanks for reporting this.  This looks like a bug in our implementation in the profiling implementation.  The situation you're describing is this:

* A user imports Proteus and starts calling Proteus functions that call logEvent (really this is a logging functionality, not profiling, but they're in the same module for historical reasons).

* Because the Profiling module hasn't been initialized, it starts off in "pre-initialization" mode, that is, log events are saved to a buffer to later be written out after MPI has been initialized.

* The buffers fill up and overflow

@cekees probably has more to say about this, but the solution I'd use is to set up the Profiling module when you launch your code.  We do this automatically for you if you use parun, but there's also an example in: https://github.com/erdc-cm/proteus/blob/master/proteus/tests/util.py#L5

You probably want to set loglevel to something much lower than 10, I think 3 or 5 is usually the right amount of verbosity.  Let me know if that helps!

In the meanwhile, let's leave this issue open.  I need to discuss with the other developers how they want to fix this issue, since the ideal solution is to call MPI initialization immediately on job launch.

**Comment by cekees** [comment]
I'm not sure what's going on here. @robertsawko can you point me to your script? 

The reason the preInitBuffer is there is so  that people can import  proteus.Profiling and and use Profiling.logEvent before communication is initialized. So I didn't plan for preInitBuffer to get very big  or to do any sort of monitoring of the buffer size. It's mainly just for the few log  messages that might be needed in a users application modules (like the p- and n-files). 

When you import iproteus it  should  initialize  Comm and Profiling and flush the buffer, so I'm not sure what  is happening in your script so that  Profiling.procID either doesn't get set or gets reset.

**Comment by ahmadia** [comment]
@cekees - Yeah, I see now.  Profiling.procID = comm.rank()

I wonder if that's not getting set properly somehow.

**Comment by robertsawko** [comment]
@cekees This happens on all of my python files that have import proteus.iproteus, but only if I run it with `python2 name_of_a_scripty.py`. If I do `parun` it gets through. Is is possible I am the only one who uses proteus without parun? Also I am getting the warning which is part of a problem too. But in any case it looks to me that the lines I put above create an infinite loop under certain conditions (warning and procID=None).

I also validated that some commits ago my procID was 0 rather than None.

**Comment by cekees** [comment]
No, many of the regressions tests use iproteus, among other things. I think
you've uncovered a very recent regression.  The tests  are now broken on my
laptop. I'm checking to see what is happening on the buildbots.

**Comment by cekees** [comment]
OK, this broke after commit 5cb363ffda6175a4bcb1d62d946812ce78d79d56.
@robertsawko, I think you should be able to verify that your scripts work
for that commit (i.e. do 'git
checkout 5cb363ffda6175a4bcb1d62d946812ce78d79d56' in proteus and then
rebuild proteus as you usually do when updating it. I will see if I can
track down the bug.

**Comment by cekees** [comment]
This broke when we merged #195 @ahmadia. When you changed the print statement to a log statement it uncovered a case in iproteus where we can create an invite loop with logEvent. The buildbots caught the error (and make check catches it on my machine), but for some reason the shippable testing didn't seem to catch it. @robertsawko , I'll push a fix shortly.

**Comment by cekees** [comment]
@ahmadia and @mfarthin. I think the reason this regression wasn't caught by shippable is that it runs the tests using nosetests, which doesn't isolate the test_* modules, so for example Profiling.procID was  getting initialized  before it hit our tests  that use iproteus. There is a  `--with-isolation` flag you can pass  to nosetests, which triggers this problem, but it also breaks a lot of other things including the coverage checking and some of @ahmadia's new tests.
